pkg/proxy/filter_endpoints.go

```diff
@@ -23,75 +23,180 @@ import (
+		// contains only Ready endpoints, so it must be a subset of clusterEndpoints.
+		allEndpoints = clusterEndpoints
+
+	default:
+		// It is possible that some endpoints appear only in internalEndpoints and
+		// others only in externalEndpoints, so we have to merge them together.
+		endpointsMap := map[string]Endpoint{}
+		for _, ep := range internalEndpoints {
+			endpointsMap[ep.String()] = ep
+		}
+		for _, ep := range externalEndpoints {
+			endpointsMap[ep.String()] = ep
+		}
+		allEndpoints = make([]Endpoint, 0, len(endpointsMap))
+		for _, ep := range endpointsMap {
+			allEndpoints = append(allEndpoints, ep)
+		}
 	}
 
-	return endpoints
+	return
 }
 
-// filterEndpointsWithHints provides filtering based on the hints included in
-// EndpointSlices. If any of the following are true, the full list of endpoints
-// will be returned without any filtering:
-// * The AnnotationTopologyAwareHints annotation is not set to "Auto" for this
-//   Service.
-// * No zone is specified in node labels.
-// * No endpoints for this Service have a hint pointing to the zone this
-//   instance of kube-proxy is running in.
-// * One or more endpoints for this Service do not have hints specified.
-func filterEndpointsWithHints(endpoints []Endpoint, hintsAnnotation string, nodeLabels map[string]string) []Endpoint {
+// canUseTopology returns true if topology aware routing is enabled and properly configured
+// in this cluster. That is, it checks that:
+// * The TopologyAwareHints feature is enabled
+// * The "service.kubernetes.io/topology-aware-hints" annotation on this Service is set to "Auto"
+// * The node's labels include "topology.kubernetes.io/zone"
+// * All of the endpoints for this Service have a topology hint
+// * At least one endpoint for this Service is hinted for this node's zone.
+func canUseTopology(endpoints []Endpoint, svcInfo ServicePort, nodeLabels map[string]string) bool {
+	hintsAnnotation := svcInfo.HintsAnnotation()
 	if hintsAnnotation != "Auto" && hintsAnnotation != "auto" {
 		if hintsAnnotation != "" && hintsAnnotation != "Disabled" && hintsAnnotation != "disabled" {
 			klog.InfoS("Skipping topology aware endpoint filtering since Service has unexpected value", "annotationTopologyAwareHints", v1.AnnotationTopologyAwareHints, "hints", hintsAnnotation)
 		}
-		return endpoints
+		return false
 	}
 
 	zone, ok := nodeLabels[v1.LabelTopologyZone]
 	if !ok || zone == "" {
 		klog.InfoS("Skipping topology aware endpoint filtering since node is missing label", "label", v1.LabelTopologyZone)
-		return endpoints
+		return false
 	}
 
-	filteredEndpoints := []Endpoint{}
-
+	hasEndpointForZone := false
 	for _, endpoint := range endpoints {
 		if endpoint.GetZoneHints().Len() == 0 {
 			klog.InfoS("Skipping topology aware endpoint filtering since one or more endpoints is missing a zone hint")
-			return endpoints
+			return false
 		}
+
+		// FIXME FIXME FIXME
+		// Should this check endpoint.IsReady()?
+		// FIXME FIXME FIXME
```

Comment:
@robscott The current rule for deciding whether to use TopologyAwareHints does "only use topology if there is at least one endpoint for this zone", but it should probably be "only use topology if there is at least one *Ready* endpoint for this zone", right?

Reply:
Yes, thanks! I'll work on a separate PR to fix that so it can be cherry picked back.

Reply:
https://github.com/kubernetes/kubernetes/pull/106507